### PR TITLE
SHS-5775: Implementation: Social Media Footer Block

### PR DESCRIPTION
**Pull request by codechefmarc**
## Summary
Adds config and update hook to create new social media footer block.

## Need Review By (Date)
2024-09-09

## Urgency
low

## Steps to Test
1. Login to the site as a Developer role
2. Under the Content menu, hover over "Blocks", then "Add content block", then click "Social Media Footer Block"

Screenshot 2024-09-05 at 1 19 07 PM

4. Fill out the fields for description, icon size, layout, and add as many links as you wish
5. Click "Save and configure" at the bottom
6. When on the "Configure block" screen, select "Footer" as the value for the "Region" field
7. At the bottom, click "Save block"
8. View the front-end of the site and scroll all the way to the footer area
9. Verify that there is a new social media footer block present, with the correct styles according to the selected icon size and layout

Grid layout:
Screenshot 2024-09-23 at 10 07 15 AM

List layout:
Screenshot 2024-09-23 at 10 07 48 AM

10. Masquerade as a user who does not have a developer role, for example someone who only has "Site manager" role
11. View the front-end of the site and scroll all the way to the footer area
12. Hover over the social media block and click the pencil icon
13. Verify that the only thing a non-developer role can do is edit the block, not remove it or configure it.

Screenshot 2024-09-23 at 10 08 54 AM

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Humsci Basic PR Checklist](https://github.com/SU-HSDO/suhumsci/blob/develop/docs/HumsciBasicPRChecklist.md)

### CODE REVIEW

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by codechefmarc** [comment]
@cienvaras - I've added the template suggestion so this is now ready for your review.

**Review by cienvaras — APPROVED** [review]
@codechefmarc Works as expected, thanks!

@ahughes3 Ready for you to test.

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by cienvaras** [comment]
@ahughes3 This one is ready to review, includes both backend ([SHS-5675](https://app.clickup.com/t/36718269/SHS-5775)) and frontend ([SHS-5676](https://app.clickup.com/t/36718269/SHS-5776)) work.

**Review by ahughes3 — APPROVED** [review]

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by joegl** [comment]
👀 I did a cursory review of this and looks good. I just want to take a closer look at the database update and .module hook. @codechefmarc There's a conflict in the `hs_paragraph_types.module` if you want merge in the latest `11.2.5-release` and resolve it that'd be great 👍

**Comment by cienvaras** [comment]
@joegl Merge conflicts fixed!

File: config/default/user.role.contributor.yml [hunk 1]
```diff
@@ -67,10 +68,7 @@ permissions:
   - 'create video media'
   - 'delete humsci entity'
   - 'delete own files'
-  - 'delete own image media'
-  - 'delete own video media'
   - 'dropzone upload files'
```
Comment (joegl) [inline config/default/user.role.contributor.yml]: Are all these permission changes supposed to be part of this?

Comment (ahughes3) [reply config/default/user.role.contributor.yml]: @codechefmarc @cienvaras

**Comment by joegl** [comment]
@codechefmarc I'm curious why you went with a custom block built via the UI and a new paragraph type, instead of creating a [custom block plugin](https://www.drupal.org/docs/creating-modules/creating-custom-blocks/create-a-custom-block-plugin) programmatically. Because of the large hook to format the icons based on the link URL's, it feels like a block plugin could have worked really well here and kept everything organized in a single block instance with a configuration form, instead of piecing together hooks and field/paragraph configuration. I think this could all comfortably fit into a single block plugin and would prefer that route if possible.

**Comment by codechefmarc** [comment]
> @codechefmarc I'm curious why you went with a custom block built via the UI and a new paragraph type, instead of creating a [custom block plugin](https://www.drupal.org/docs/creating-modules/creating-custom-blocks/create-a-custom-block-plugin) programmatically.

@joegl - This was discussed internally before we started this ticket and we thought it was best to create a block via the UI because of the repeating mulit-value field with the paragraph, maintainability, and the ability to more easily add fields to this block if needed via the UI. I certainly can see it both ways - there are pros and cons to building this way vs a custom block plugin.

File: config/default/user.role.contributor.yml (hunk 1 above)
Comment (codechefmarc) [reply config/default/user.role.contributor.yml]: No, those should not be there, good catch. I'll remove and resubmit.

**Comment by ahughes3** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1619-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-colorful-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://hs-traditional-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com
- https://english-1ke5pxzkt6bpvyesr1btonierml3b2ye.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce

Visual Diffs:
- https://dashboard.tugboatqa.com/66da0e4e3c7b1b68d78dc1ce/visualdiffs

**Comment by codechefmarc** [comment]
@joegl - Reverted permissions to 11.3.1-release version. Ready for a retest. Thanks!

File: config/default/user.role.contributor.yml (hunk 1 above)
Comment (codechefmarc) [reply config/default/user.role.contributor.yml]: Fixed.

**Comment by joegl** [comment]
> > @codechefmarc I'm curious why you went with a custom block built via the UI and a new paragraph type, instead of creating a [custom block plugin](https://www.drupal.org/docs/creating-modules/creating-custom-blocks/create-a-custom-block-plugin) programmatically.
> 
> @joegl - This was discussed internally before we started this ticket and we thought it was best to create a block via the UI because of the repeating mulit-value field with the paragraph, maintainability, and the ability to more easily add fields to this block if needed via the UI. I certainly can see it both ways - there are pros and cons to building this way vs a custom block plugin.

Ok, that makes sense. I think the use of a new paragraph type to capture a multi-value field for a block is a bit out of scope for the intended use of paragraphs. A basic URL field comes with a title and URL which could suffice for the label and URL being captured here. A single block plugin also would probably be easier to maintain than all the configuration and module hooks here. 

We can still move forward with this implementation and I understand the reasons you chose this route. But I do want to push again to use a custom block plugin instead if possible.

**Comment by codechefmarc** [comment]
> > > @codechefmarc I'm curious why you went with a custom block built via the UI and a new paragraph type, instead of creating a [custom block plugin](https://www.drupal.org/docs/creating-modules/creating-custom-blocks/create-a-custom-block-plugin) programmatically.
> > 
> > 
> > @joegl - This was discussed internally before we started this ticket and we thought it was best to create a block via the UI because of the repeating mulit-value field with the paragraph, maintainability, and the ability to more easily add fields to this block if needed via the UI. I certainly can see it both ways - there are pros and cons to building this way vs a custom block plugin.
> 
> Ok, that makes sense. I think the use of a new paragraph type to capture a multi-value field for a block is a bit out of scope for the intended use of paragraphs. A basic URL field comes with a title and URL which could suffice for the label and URL being captured here. A single block plugin also would probably be easier to maintain than all the configuration and module hooks here.
> 
> We can still move forward with this implementation and I understand the reasons you chose this route. But I do want to push again to use a custom block plugin instead if possible.

@joegl - I'm going to create a new PR to address this and create a custom block instead of one via the UI. For the moment, I'll keep this PR open, but will mark it as don't merge and once we're happy with the other PR, we can close this one.

**Comment by ahughes3** [comment]
@cienvaras @codechefmarc withthe other PR open, do we still need this one?

**Comment by cienvaras** [comment]
@ahughes3 That's right, closing this now.

**Pull request closed without merging**